### PR TITLE
[WIP] Static code analysis in integration build (lint cppcheck)

### LINKED ISSUE #978: [CI] Integrate lint
clang-tidy or cppcheck

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

Static code analysis in integration build.
Handle current violations.
Fixes #978 .

## Testing instructions

New CI build.

```
cmake -H. -Bbuild -DCMAKE_BUILD_TYPE=Debug -DCPP_CHECK=ON
make -Cbuild tests
```

## Types of changes

* New feature, infrastructure (non-breaking change which adds functionality)

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [ ] Update documentation as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

### CODE REVIEW

File: .github/workflows/linux-lint.yml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      with:
+        path: build/local
+        key: ${{ runner.os }}-internal-${{ hashFiles('tools/install-dependencies') }}
+    - name: Install internal dependencies
+      run: |
+        tools/install-dependencies
+      env:
+        CC: /usr/bin/clang
+        CXX: /usr/bin/clang++
+      if: steps.internal_cache.outputs.cache-hit != 'true'
```
Comment (vikmeup) [inline .github/workflows/linux-lint.yml:31]: would this lint be consistent with local?

Comment (catenocrypt) [reply .github/workflows/linux-lint.yml:31]: All linting settings are consistent, coming from Cmake and cppcheck exception files.  What could be different is the cppcheck version -- this build installs with apt-get on Ubuntu 18.04, version is 1.82, which is not the latest.

**Comment by hewigovens** [comment]
let's merge `linux-lint.yml` into `linux-ci.yml`? the setup steps can be used

**Comment by catenocrypt** [comment]
> let's merge `linux-lint.yml` into `linux-ci.yml`? the setup steps can be used

I will probably do that; I started with a separate one until check is set up properly.  Build time will increase a lot unfortunately.  I'm currently facing some errors without any error message...

**Comment by hewigovens** [comment]
Build and lint took 2h 7m 25s (only 70%), 😨

**Comment by catenocrypt** [comment]
I'm extending the ignore list, ignoring tests.  Also, multi-thread build is disabled, as I was suspecting unstable behaviour with cppcheck.  I will try to enable it, should speed up.
On a positive note, there are not many violations!

**Comment by hewigovens** [comment]
if it takes too long, we probably need to run it periodically, not every pr

**Comment by catenocrypt** [comment]
Yes.  Or *after* every merge.  Checking diff only before a merge is another option, though a bit more complex -- cannot take use of CMake infrastructure for supplying build settings like include paths.

File: .github/workflows/linux-lint.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    - name: Build and test
+      run: |
+        tools/generate-files
```
Comment (vikmeup) [inline .github/workflows/linux-lint.yml:34]: can you separate them out into Build, Test

Comment (catenocrypt) [reply .github/workflows/linux-lint.yml:34]: As in this setup CMake drives cppcheck, cppcheck without build is not possible.

**Comment by catenocrypt** [comment]
Cppcheck is very slow, trying clang-tidy

**Pull request closed without merging**